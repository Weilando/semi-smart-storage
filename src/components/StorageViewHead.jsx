import React from 'react';

function StorageViewHead(props) {
  return (
    <thead>
      <tr>
        <th>Name</th>
        <th>Unit</th>
        <th>Entity</th>
        <th>Options</th>
      </tr>
    </thead>
  );
}

export default StorageViewHead;
