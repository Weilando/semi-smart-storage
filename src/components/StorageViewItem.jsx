import React from 'react';
import PropTypes from 'prop-types';

function StorageViewItem(props) {
  StorageViewItem.proptTypes = {
    iId: PropTypes.number.isRequired,
    iName: PropTypes.string.isRequired,
    iUnit: PropTypes.string.isRequired,
    iQuantity: PropTypes.number.isRequired,
    decrementAction: PropTypes.func.isRequired,
    incrementAction: PropTypes.func.isRequired,
    editAction: PropTypes.func.isRequired,
    removeAction: PropTypes.func.isRequired,
  };

  return(
    <tr>
      <td>{props.iName}</td>
      <td>{props.iUnit}</td>
      <td>{props.iQuantity}</td>
      <td>
        <button className='StorageView' onClick={() => props.decrementAction(props.iId)}>-</button>
        <button className='StorageView' onClick={() => props.incrementAction(props.iId)}>+</button>
        <button className='StorageView' onClick={() => props.editAction(props.iId)}>Edit</button>
        <button className='StorageView' onClick={() => props.removeAction(props.iId)}>Remove</button>
      </td>
    </tr>
  );
}

export default StorageViewItem;
