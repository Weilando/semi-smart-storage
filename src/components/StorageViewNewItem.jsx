import React from 'react';
import PropTypes from 'prop-types';
import { AddMode } from '../constants/enums';
import '../styles/StorageView.css';

class StorageViewNewItem extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      item: -1,
      unit: -1,
      quantity: 0,
    }

    this.handleItemChange = this.handleItemChange.bind(this);
    this.handleUnitChange = this.handleUnitChange.bind(this);
    this.handleQuantityChange = this.handleQuantityChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
  }

  static propTypes = {
    storageId: PropTypes.number.isRequired,
    itemList: PropTypes.arrayOf(
      PropTypes.shape(
        {id: PropTypes.number.isRequired, name: PropTypes.string.isRequired}
      )
    ).isRequired,
    unitList: PropTypes.arrayOf(
      PropTypes.shape(
        {id: PropTypes.number.isRequired, name: PropTypes.string.isRequired}
      )
    ).isRequired,
    addAction: PropTypes.func.isRequired,
  }


  // handling inputs
  handleItemChange(event) {
    this.setState({item: Number(event.target.value)});
  }

  handleUnitChange(event) {
    this.setState({unit: Number(event.target.value)});
  }

  handleQuantityChange(event) {
    this.setState({quantity: Number(event.target.value)});
  }

  handleSubmit(event) {
    event.preventDefault();
    this.props.addAction(AddMode.STORAGE_CONTENT, this.state.item, this.state.unit, this.state.quantity, this.props.storageId);
    this.setState({item: -1, unit: -1, quantity: 0});
  }

  // rendering
  unpackItemList() {
    return this.props.itemList.map((currItem) =>
      <option key={currItem.id} value={currItem.id}>
        {currItem.name}
      </option>
    );
  }

  unpackUnitList() {
    return this.props.unitList.map((currUnit) =>
      <option key={currUnit.id} value={currUnit.id}>
        {currUnit.name}
      </option>
    );
  }

  render() {
    let itemOptions = this.unpackItemList();
    let unitOptions = this.unpackUnitList();

    return(
      <tfoot>
        <tr>
          <td>
            <select id="Item" name="Item" onChange={this.handleItemChange} value={this.state.item}>
              <option className="choose" value={-1}>-Choose item-</option>
              {itemOptions}
            </select>
          </td>
          <td>
            <select id="Unit" name="Unit" onChange={this.handleUnitChange} value={this.state.unit}>
              <option className="choose" value={-1}>-Choose unit-</option>
              {unitOptions}
            </select>
          </td>
          <td colSpan="2">
            <form onSubmit={this.handleSubmit}>
              <input type="number"
                min="0" step="0.05"
                value={this.state.quantity}
                onChange={this.handleQuantityChange}
              />
              <input type="submit" value="Add"
                disabled={(this.state.quantity === '') || (this.state.item===-1) || (this.state.unit===-1)}
              />
            </form>
          </td>
        </tr>
      </tfoot>
    );
  }
}

export default StorageViewNewItem;
