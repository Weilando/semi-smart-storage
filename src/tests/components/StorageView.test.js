import React from 'react';
import renderer from 'react-test-renderer';
import StorageViewHead from '../../components/StorageViewHead';
import StorageViewItem from '../../components/StorageViewItem';
import StorageViewNewItem from '../../components/StorageViewNewItem';

describe('snapshot tests for StorageViewHead component', () => {
  it('should render StorageViewHead component correctly', () => {
    const tree = renderer.create(
      <StorageViewHead/>
    ).toJSON();

    expect(tree).toMatchSnapshot();
  })
});

describe('snapshot tests for StorageViewItem component', () => {
  it('should render StorageViewItem component correctly', () => {
    const editAction = jest.fn();
    const decrementAction = jest.fn();
    const incrementAction = jest.fn();
    const removeAction = jest.fn();

    const tree = renderer.create(
      <StorageViewItem
        iId={42}
        iName="Milk"
        iUnit="1L"
        iQuantity={2.5}
        decrementAction={decrementAction}
        incrementAction={incrementAction}
        editAction={editAction}
        removeAction={removeAction}
      />
    ).toJSON();

    expect(tree).toMatchSnapshot();
  })
});

describe('snapshot tests for StorageViewNewItem component', () => {
  it('should render empty StorageViewNewItem component correctly', () => {
    const addAction = jest.fn();
    const items = [];
    const units = [];

    const tree = renderer.create(
      <StorageViewNewItem
        addAction={addAction}
        itemList={items}
        unitList={units}
      />
    ).toJSON();

    expect(tree).toMatchSnapshot();
  })

  it('should render StorageViewNewItem component with units and items correctly', () => {
    const addAction = jest.fn();
    const items = [{id: 0, name: "Sparkling water"}, {id: 1, name: "Apple juice"}, {id: 2, name: "Craft beer"}, {id: 3, name: "Milk"}, {id: 5, name: "Coke"}];
    const units = [{id: 0, name: "0.33L"}, {id: 1, name: "0.5L"}, {id: 2, name: "1.0L"}];

    const tree = renderer.create(
      <StorageViewNewItem
        addAction={addAction}
        itemList={items}
        unitList={units}
      />
    ).toJSON();

    expect(tree).toMatchSnapshot();
  })
});
