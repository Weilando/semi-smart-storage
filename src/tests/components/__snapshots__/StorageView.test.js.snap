// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot tests for StorageViewHead component should render StorageViewHead component correctly 1`] = `
<thead>
  <tr>
    <th>
      Name
    </th>
    <th>
      Unit
    </th>
    <th>
      Entity
    </th>
    <th>
      Options
    </th>
  </tr>
</thead>
`;

exports[`snapshot tests for StorageViewItem component should render StorageViewItem component correctly 1`] = `
<tr>
  <td>
    Milk
  </td>
  <td>
    1L
  </td>
  <td>
    2.5
  </td>
  <td>
    <button
      className="StorageView"
      onClick={[Function]}
    >
      -
    </button>
    <button
      className="StorageView"
      onClick={[Function]}
    >
      +
    </button>
    <button
      className="StorageView"
      onClick={[Function]}
    >
      Edit
    </button>
    <button
      className="StorageView"
      onClick={[Function]}
    >
      Remove
    </button>
  </td>
</tr>
`;

exports[`snapshot tests for StorageViewNewItem component should render StorageViewNewItem component with units and items correctly 1`] = `
<tfoot>
  <tr>
    <td>
      <select
        id="Item"
        name="Item"
        onChange={[Function]}
      >
        <option
          value={0}
        >
          Sparkling water
        </option>
        <option
          value={1}
        >
          Apple juice
        </option>
        <option
          value={2}
        >
          Craft beer
        </option>
        <option
          value={3}
        >
          Milk
        </option>
        <option
          value={5}
        >
          Coke
        </option>
      </select>
    </td>
    <td>
      <select
        id="Unit"
        name="Unit"
        onChange={[Function]}
      >
        <option
          value={0}
        >
          0.33L
        </option>
        <option
          value={1}
        >
          0.5L
        </option>
        <option
          value={2}
        >
          1.0L
        </option>
      </select>
    </td>
    <td>
      <input
        min="0"
        onChange={[Function]}
        type="number"
        value={0}
      />
    </td>
    <td>
      <button
        disabled={false}
        onClick={[Function]}
      >
        Add
      </button>
    </td>
  </tr>
</tfoot>
`;

exports[`snapshot tests for StorageViewNewItem component should render empty StorageViewNewItem component correctly 1`] = `
<tfoot>
  <tr>
    <td>
      <select
        id="Item"
        name="Item"
        onChange={[Function]}
      />
    </td>
    <td>
      <select
        id="Unit"
        name="Unit"
        onChange={[Function]}
      />
    </td>
    <td>
      <input
        min="0"
        onChange={[Function]}
        type="number"
        value={0}
      />
    </td>
    <td>
      <button
        disabled={false}
        onClick={[Function]}
      >
        Add
      </button>
    </td>
  </tr>
</tfoot>
`;
